src/peft/utils/loftq_utils.py

```diff
@@ -227,3 +233,139 @@ def loftq_init(weight: Union[torch.Tensor, torch.nn.Parameter], num_bits: int, r
     lora_A, lora_B = R, L
 
     return dequantized_weight.to(device=device, dtype=dtype), lora_A, lora_B
+
+
+def _check_model_path_loftq(model_path, peft_model):
+    """Check if there is a local model file corresponding to the given model, then format it properly."""
+    if model_path is None:
+        try:
+            model_path = snapshot_download(peft_model.base_model.config._name_or_path, local_files_only=True)
+        except AttributeError as exc:
+            raise ValueError(
+                "The provided model does not appear to be a transformers model. In this case, you must pass the "
+                "model_path to the safetensors file."
+            ) from exc
+        except LocalEntryNotFoundError as exc:
+            raise ValueError("The model.safetensors file must be present on disk, but it could not be found.") from exc
+
+    suffix = "model.safetensors"
+    if not model_path.endswith(suffix):
+        model_path = os.path.join(model_path, suffix)
+    if not os.path.exists(model_path):
+        raise FileNotFoundError(
+            f"Could not find file {model_path}, ensure that there is a safetensors file of the model."
+        )
+
+    return model_path
+
+
+@torch.no_grad()
+def _loftq_init_new(qweight, weight, num_bits: int, reduced_rank: int):
```

Comment:
Asd per the loftq algo, there should be alternating steps to quantize Qt = W-B(t-1)A(t-1) and At,Bt ← SVD(W − Qt) for T steps. here, we only do a single step of A1,B1 ← SVD(W − Qbnb)

![Screenshot 2024-03-19 at 12 36 57 PM](https://github.com/huggingface/peft/assets/13534540/f2fc0cb7-588c-4aa7-99e2-f3a4c6bbb428)

Reply:
As discussed internally, since this approach intends not to adapt the quantized weights (to avoid having an extra copy of all these weights), only single step is implemented.